registry/storage/driver/s3-aws/s3.go

```diff
@@ -787,50 +789,52 @@ func (d *driver) statList(ctx context.Context, s3Path string) (*storagedriver.Fi
 	if err != nil {
 		return nil, err
 	}
-	var fi storagedriver.FileInfoFields
 	if len(resp.Contents) == 1 {
 		if *resp.Contents[0].Key != s3Path {
-			fi.IsDir = true
-			return &fi, nil
+			return &storagedriver.FileInfoFields{
+				Path:  path,
+				IsDir: true,
+			}, nil
 		}
-		fi.IsDir = false
-		fi.Size = *resp.Contents[0].Size
-		fi.ModTime = *resp.Contents[0].LastModified
-		return &fi, nil
+		return &storagedriver.FileInfoFields{
+			Path:    path,
+			Size:    *resp.Contents[0].Size,
+			ModTime: *resp.Contents[0].LastModified,
+		}, nil
 	}
 	if len(resp.CommonPrefixes) == 1 {
-		fi.IsDir = true
-		return &fi, nil
+		return &storagedriver.FileInfoFields{
+			Path:  path,
+			IsDir: true,
+		}, nil
 	}
 	return nil, storagedriver.PathNotFoundError{Path: s3Path}
 }
 
 // Stat retrieves the FileInfo for the given path, including the current size
 // in bytes and the creation time.
 func (d *driver) Stat(ctx context.Context, path string) (storagedriver.FileInfo, error) {
-	s3Path := d.s3Path(path)
-	fi, err := d.statHead(ctx, s3Path)
+	fi, err := d.statHead(ctx, path)
 	if err != nil {
 		// For AWS errors, we fail over to ListObjects:
 		// Though the official docs https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html#API_HeadObject_Errors
 		// are slightly outdated, the HeadObject actually returns NotFound error
 		// if querying a key which doesn't exist or a key which has nested keys
 		// and Forbidden if IAM/ACL permissions do not allow Head but allow List.
-		if _, ok := err.(awserr.Error); ok {
-			fi, err := d.statList(ctx, s3Path)
+		var awsErr awserr.Error
+		if errors.As(err, &awsErr) {
+			fi, err := d.statList(ctx, path)
 			if err != nil {
 				if errors.Is(err, storagedriver.PathNotFoundError{}) {
 					return nil, storagedriver.PathNotFoundError{Path: path}
 				}
 				return nil, parseError(path, err)
```

Comment:
Actually, looks like this can also be reduces now that `statList` gets the path passed.

Reply:
Looks like this code doesn't work;

- `errors.Is()` only matches if a `storagedriver.PathNotFoundError{}` is returned _without_ a `Path` set, but `statList` returns `storagedriver.PathNotFoundError{Path: s3Path}`, so it would never be empty
- ☝️ that should probably be `errors.As` to detect the type
- This code replaces the error (containing `s3Path`) and then changes it to use an error with `path` instead.

Reply:
`errors.Is()` breaking on `storagedriver.PathNotFoundError` is interesting; I find it kinda funny because originally I was actually returning empty `storagedriver.PathNotFoundError` but then decided to smack `s3Path` into it -- that value is actually overridden in `Stat` call but I admit this is an oversight on my side; we should use `errors.As` there

Reply:
I hate the errors.Is/errors.As. Or at least; they're too easy to get wrong, and nothing will immediately reveal that you did so.